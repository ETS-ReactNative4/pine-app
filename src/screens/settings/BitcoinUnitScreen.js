import React, { Component } from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { save as saveSettings } from '../../actions/settings/save';
import headerStyles from '../../styles/headerStyles';
import settingsStyles from '../../styles/settingsStyles';
import BackButton from '../../components/BackButton';
import SettingsGroup from '../../components/SettingsGroup';
import SettingsOption from '../../components/SettingsOption';
import SettingsDescription from '../../components/SettingsDescription';
import SettingsTitle from '../../components/SettingsTitle';
import StrongText from '../../components/StrongText';
import BtcLabel from '../../components/BtcLabel';
import BaseSettingsScreen from './BaseSettingsScreen';
import config from '../../config';

@connect((state) => ({
  settings: state.settings,
  balance: state.bitcoin.wallet.balance
}))
export default class BitcoinUnitScreen extends Component {
  static navigationOptions = ({ navigation }) => ({
    title: 'Bitcoin Display Unit',
    headerStyle: headerStyles.header,
    headerTitleStyle: headerStyles.title,
    headerLeft: (<BackButton onPress={() => { navigation.goBack(); }} />)
  });

  constructor(props) {
    super(...arguments);

    this.state = {
      unit: props.settings.bitcoin.unit
    };
  }

  componentWillUnmount() {
    this._save();
  }

  _save() {
    const dispatch = this.props.dispatch;
    const unit = this.state.unit || config.bitcoin.unit;

    dispatch(saveSettings({
      bitcoin: { unit }
    }));
  }

  _onSelect(unit) {
    this.setState({ unit });
  }

  render() {
    return (
      <BaseSettingsScreen>
        <SettingsTitle>Balance Preview</SettingsTitle>
        <SettingsGroup>
          <View style={[settingsStyles.item, { borderBottomWidth: 0, alignItems: 'center' }]}>
            <BtcLabel amount={this.props.balance} unit={this.state.unit} style={settingsStyles.label} />
          </View>
        </SettingsGroup>
        <SettingsDescription>
          This is how your current wallet balance will look like with the selected unit below.
        </SettingsDescription>

        <SettingsTitle>Display Unit</SettingsTitle>
        <SettingsGroup>
          <SettingsOption name='BTC' value={this.state.unit} onSelect={this._onSelect.bind(this)} />
          <SettingsOption name='mBTC' value={this.state.unit} onSelect={this._onSelect.bind(this)} />
          <SettingsOption name='Satoshis' value={this.state.unit} onSelect={this._onSelect.bind(this)} isLastItem={true} />
        </SettingsGroup>

        <SettingsDescription>
          Choose the unit in which bitcoin amounts should be displayed.
        </SettingsDescription>
        <SettingsDescription />
        <SettingsDescription>
          <StrongText>BTC</StrongText> is one bitcoin and will be displayed as BTC.
        </SettingsDescription>
        <SettingsDescription>
          <StrongText>mBTC</StrongText> is one thousandth of a bitcoin and will be displayed as mBTC.
        </SettingsDescription>
        <SettingsDescription>
          <StrongText>Satoshi</StrongText> is one hundred millionth of a bitcoin and will be displayed as sats.
        </SettingsDescription>
      </BaseSettingsScreen>
    );
  }
}

BitcoinUnitScreen.propTypes = {
  settings: PropTypes.object,
  balance: PropTypes.number,
  dispatch: PropTypes.func,
  navigation: PropTypes.any
};
