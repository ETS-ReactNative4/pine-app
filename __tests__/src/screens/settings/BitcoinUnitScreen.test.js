import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import BitcoinUnitScreen from '../../../../src/screens/settings/BitcoinUnitScreen';

jest.mock('../../../../src/containers/BalanceLabelContainer', () => 'BalanceLabelContainer');

const storeMock = {
  getState: jest.fn(() => ({
    settings: {
      bitcoin: {
        unit: 'BTC'
      }
    }
  })),
  dispatch: jest.fn(),
  subscribe: jest.fn()
};

const navigationMock = {
  navigate: jest.fn(),
  state: {
    params: {}
  }
};

describe('BitcoinUnitScreen', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <BitcoinUnitScreen store={storeMock} navigation={navigationMock} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
