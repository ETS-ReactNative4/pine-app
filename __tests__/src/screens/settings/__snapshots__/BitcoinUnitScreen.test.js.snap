// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitcoinUnitScreen renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#EFEFF3",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Array [
            Object {
              "paddingTop": 35,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 5.5,
                "paddingLeft": 15,
                "paddingRight": 15,
                "textTransform": "uppercase",
              },
            ]
          }
        >
          Balance Preview
        </Text>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "borderBottomColor": "#C8C8CC",
                "borderBottomWidth": 0.5,
                "borderTopColor": "#C8C8CC",
                "borderTopWidth": 0.5,
                "marginBottom": 35,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderBottomColor": "#C8C7CC",
                  "borderBottomWidth": 0.5,
                  "flex": 1,
                  "height": 44,
                  "justifyContent": "center",
                  "marginLeft": 15,
                  "paddingRight": 15,
                },
                Object {
                  "alignItems": "center",
                  "borderBottomWidth": 0,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "fontSize": 17,
                    "fontWeight": "400",
                    "lineHeight": 22,
                  },
                ]
              }
            >
              1.37484 BTC
            </Text>
          </View>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        >
          This is how your current account balance looks like with the selected unit below.
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 5.5,
                "paddingLeft": 15,
                "paddingRight": 15,
                "textTransform": "uppercase",
              },
            ]
          }
        >
          Display Unit
        </Text>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "borderBottomColor": "#C8C8CC",
                "borderBottomWidth": 0.5,
                "borderTopColor": "#C8C8CC",
                "borderTopWidth": 0.5,
                "marginBottom": 35,
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "borderBottomColor": "#C8C7CC",
                    "borderBottomWidth": 0.5,
                    "flex": 1,
                    "height": 44,
                    "justifyContent": "center",
                    "marginLeft": 15,
                    "paddingRight": 15,
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "fontSize": 17,
                      "fontWeight": "400",
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                BTC
              </Text>
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    Object {
                      "color": "#007AFF",
                      "fontSize": 16,
                      "paddingTop": 2,
                      "position": "absolute",
                      "right": 15,
                    },
                    Object {
                      "fontFamily": "Octicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                ï„”
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "borderBottomColor": "#C8C7CC",
                    "borderBottomWidth": 0.5,
                    "flex": 1,
                    "height": 44,
                    "justifyContent": "center",
                    "marginLeft": 15,
                    "paddingRight": 15,
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "fontSize": 17,
                      "fontWeight": "400",
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                mBTC
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "borderBottomColor": "#C8C7CC",
                    "borderBottomWidth": 0.5,
                    "flex": 1,
                    "height": 44,
                    "justifyContent": "center",
                    "marginLeft": 15,
                    "paddingRight": 15,
                  },
                  Object {
                    "borderBottomWidth": 0,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "fontSize": 17,
                      "fontWeight": "400",
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                Satoshis
              </Text>
            </View>
          </View>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        >
          Choose the unit in which bitcoin amounts should be displayed.
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "fontWeight": "600",
              }
            }
          >
            BTC
          </Text>
           is one bitcoin and will be displayed as BTC.
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "fontWeight": "600",
              }
            }
          >
            mBTC
          </Text>
           is one thousandth of a bitcoin and will be displayed as mBTC.
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "fontWeight": "600",
              }
            }
          >
            Satoshi
          </Text>
           is one hundred millionth of a bitcoin and will be displayed as sats.
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
